Reimplementation.

jaheba opened the pull request:
work in progress

jaheba committed:
Reimplementation.

jaheba committed:
Formatting (cargo fmt).

vext01 commented:
Mind if I add some tests?

vext01 commented on src/lib.rs:
```diff
@@ -1,108 +1,132 @@
-            nodes: nodes,
+impl<T> NodeData<T> {
+    fn new(val: T,
+           tree: *const RefCell<PPTree<T>>,
```
Why do you need the pointer?

vext01 commented on src/lib.rs:
```diff
@@ -1,108 +1,132 @@
-
-pub struct NodeIter<'a, T: 'a> {
-    node: Option<&'a Node<T>>,
+struct PPTree<T> {
```
Is there any need to have this struct? I think the whole PPT could be defined only in terms of nodes? Unless you need to know, for example, the number of nodes in a whole tree...

vext01 reviewed (commented):
I've added a couple of comments, but before I look further, can you add some comments? For example, what should push (intuitively) do?

I've not been able to find a good description of the operations available on a PPT. Did you find one?

(If you wanted some Wikipedia Karma, you could add a description of the operations to the wiki page)

Thanks

jaheba committed:
Renaming. Added RStack.

jaheba commented:
I've moved and renamed a lot of things. There are now two implementations. The one in `rec` should be more straightforward. Currently working on documentation.

jaheba committed:
Added some documentation.

jaheba committed:
Updated repository.

ltratt commented on src/lib.rs:
```diff
@@ -1,108 +1,92 @@
+//!
+//! # Overview
+//!
+//! The stacks described in this crate differ from traditional stacks in one
```
Conceptually this is an immutable linked list, which Lispy languages (and many others) call "cons cells". Maybe it might be worth thinking about using that terminology?

ltratt commented on src/lib.rs:
````diff
@@ -1,108 +1,92 @@
+//! assert_eq!(*two, 2);
+//! ```
+//! Further, popping a value from the stack just returns the parent -- the
+//! originial value (and thus the stack it represents) remains valid:
````
Typo: "original"

ltratt commented on src/lib.rs:
```diff
@@ -1,108 +1,92 @@
+//!
+//! ## Cactus stacks
+//!
+//! Due to the immutable property, it is possible to spawn of multiple values
```
Typo: "off".

ltratt commented on src/rec.rs:
```diff
@@ -0,0 +1,76 @@
+// }
+
+// #[test]
+// fn test() {
```
Why's the test code commented out?

ltratt commented on src/vec.rs:
```diff
@@ -0,0 +1,125 @@
+
+
+impl<T> VStack<T> {
+    fn deref_node(&self) -> &Node<T> {
```
I *think* I can break this with something like:

```rust
let t = Tree{...};
let n = Node{tree: RefCell::new(t)};
let v = VStack{tree: t};
let n2 = v.deref_node(); // An immutable reference wrapping t
n.push(...); // Mutates t, so what happens to n2?
```

Warning: I might be very wrong.

vext01 commented:
(Perhaps in my naivety) I think we can make something much simpler. I'm going to have a play this afternoon.

jaheba committed:
Experimented with vec solution.

jaheba committed:
Removed vec implementation.

jaheba committed:
Typo; Stack cannot be empty

jaheba committed:
Bump version.

jaheba committed:
Added iterator.

vext01 commented:
What's the status of this PR Jasper?

jaheba committed:
Reworked Stack implementation (again).

There is a trait `PushPop` which defines basic operation on stacks.
`Cell` is a cons cell. `Stack` a newtype for `Rc<Cell>`.

`PushPop` is implemented for `Stack` and `Option<Stack>`. A Stack
always contains value. Empty stacks are represented by `Option<Stack>::None`.

ltratt commented:
Can we add some tests to the (re)implementation so that we can understand how it might be to use it in practise?

jaheba committed:
Added some tests.

jaheba commented:
I've added some tests. Documentation is still inaccurate though.

vext01 commented:
OK, let me know when you are ready for re-review.

jaheba committed:
Bump version.

jaheba committed:
Added support for default values.

jaheba committed:
Bump version.

vext01 commented:
Can we close this?

jaheba commented:
I think we could just close the repository in favour of sofdevteam/cactus.

ltratt commented:
I'll close this PR. It's then up to you whether you want to delete the repo or not (there isn't a "close" option other than "delete").
